src/contexts/NetworkContext.js

```diff
@@ -0,0 +1,61 @@
+import React, { createContext, useState, useEffect, useContext } from 'react';
+import storage from '../services/storage';
+
+const NetworkConext = createContext();
+
+/**
+ * Custom hook to subscribe to network status.
+ *
+ * Value is retrieved from local storage initially and updated by service worker firing events
+ *
+ * Can be used by any functional component under the NetworkContextProvider in App.js
+ *
+ * @returns {boolean} true if connected to the network, false otherwise.
+ */
+export const useNetworkStatus = () => {
+  const context = useContext(NetworkConext);
+  if (context === undefined) {
+    throw new Error(`useNetworkStatus must be called within NetworkContextProvider`);
+  }
+
+  return context;
+};
+
+const NetworkContextProvider = ({ children }) => {
+  const [isOnline, setIsOnline] = useState('online');
+
+  useEffect(() => {
+    // Retrieve network status from local storage or default to online
+    try {
+      const networkStatus = storage.get('network-status');
+      setIsOnline(networkStatus === 'online');
+    } catch (error) {
+      console.error(error);
+      setIsOnline(true);
+    }
+
+    /* Used to re-render the page when the network connection changes.
+     * The origin of the message is checked to prevent cross-site scripting attacks
+     */
+    const updateNetworkStatus = (event) => {
+      setIsOnline((prevStatus) => {
+        if (
+          event.origin === window.location.origin &&
+          (event.data === 'online' || event.data === 'offline')
```

Comment:
Trying to understand this security check: Are we checking the event.data to make sure we are not getting any data other than those exact two states?

Reply:
If you do not mind explaining, too, is event.data basically supposed to mirror prevStatus? Or, if event.data is an indication of what the status should be changed to, why do we change it to online whether event.data is online OR offline?

Reply:
The security aspect is mostly in the first half of the boolean, `event.origin === window.location.origin`. This ensures that we aren't accepting events from external sites (which can have arbitrary - aka malicious - code inside them).

Reply:
`event.data` is a property of the event set by whatever code created the event. In this case, we're expecting the event to originate from our serviceWorker. The serviceWorker will send out an event of type `message` whenever the network connection status changes. This status is (I believe by browser API standards) defined as either 'online' or 'offline'. So this function is getting called for all `message` events (see my below comment for further details on events), and ensuring that the event is originating from within our code. It is also checking that the event contains data relevant to the network connection status. The serviceWorker could send `message` events about several other things that aren't relevant to this component, so we check if the data is one of the network-statuses to filter out the irrelevant events.

If you look again at the return when the event is legitimate and relevant (line 28), you'll notice that it's actually returning a `boolean` value, which is whether the event's data property is equal to 'online'. Since we already know the data is one of either 'online' or 'offline', this is equivalent to: 

```
if (event.data === 'online') {
    return true
} else if (event.data === 'offline') {
    return false
}
```